staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1019 @@
+	return func(options *k8smetav1.ListOptions) {
+		// TODO if a selector is both in options.FieldSelector and customFieldsSelector it appears twice in the
+		// resulting selector. This is not incorrect but it's redundant (and less efficient I suspect).
+		// Fix this: if a selector appears both in options.FieldSelector and customFieldSelector it must
+		// appear in the resulting selector only once.
+		optionsFieldSelector := options.FieldSelector
+		allSelectors := make([]string, 0, 2)
+		if strings.Trim(optionsFieldSelector, " ") != "" {
+			allSelectors = append(allSelectors, optionsFieldSelector)
+		}
+		allSelectors = append(allSelectors, customFieldSelector)
+		options.FieldSelector = strings.Join(allSelectors, ",")
+	}
+}
+
+// attachmentIfcName is an Index function that returns
+// the status.IfcName of a NetworkAttachment.
+func attachmentIfcName(obj interface{}) ([]string, error) {
+	att := obj.(*netv1a1.NetworkAttachment)
+	return []string{att.Status.IfcName}, nil
+}
+
+func v1a1AttsCustomInformerAndLister(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) (k8scache.SharedIndexInformer, koslisterv1a1.NetworkAttachmentLister) {
+
+	attv1a1Informer := createAttsv1a1Informer(kcs,
+		resyncPeriod,
+		k8smetav1.NamespaceAll,
+		tweakListOptionsFunc)
+	return attv1a1Informer.Informer(), attv1a1Informer.Lister()
+}
+
+func v1a1AttsCustomNamespaceInformerAndLister(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	namespace string,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) (k8scache.SharedIndexInformer, koslisterv1a1.NetworkAttachmentNamespaceLister) {
+
+	attv1a1Informer := createAttsv1a1Informer(kcs,
+		resyncPeriod,
+		namespace,
+		tweakListOptionsFunc)
+	return attv1a1Informer.Informer(), attv1a1Informer.Lister().NetworkAttachments(namespace)
+}
+
+func createAttsv1a1Informer(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	namespace string,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) kosinformersv1a1.NetworkAttachmentInformer {
+
+	localAttsInformerFactory := kosinformers.NewFilteredSharedInformerFactory(kcs,
+		resyncPeriod,
+		namespace,
+		tweakListOptionsFunc)
+	netv1a1Ifc := localAttsInformerFactory.Network().V1alpha1()
+	return netv1a1Ifc.NetworkAttachments()
+}
+
+// TODO factor out attNSN in pkg controller/utils, as it is used both by
+// the connection agent and the IPAM controller.
+func attNSN(netAtt *netv1a1.NetworkAttachment) k8stypes.NamespacedName {
+	return k8stypes.NamespacedName{Namespace: netAtt.Namespace,
+		Name: netAtt.Name}
+}
+
+func fromAttRefToVNIAndNsn(attRef attQueueRef) vniAndNsn {
+	return vniAndNsn{
+		attRef.vni,
+		attRef.nsn,
+	}
+}
+
+func fromAttToVNIAndNsn(att *netv1a1.NetworkAttachment) vniAndNsn {
+	return vniAndNsn{
+		att.Spec.VNI,
+		attNSN(att),
+	}
+}
+
+func generateMACAddr(vni uint32, guestIPv4 gonet.IP) gonet.HardwareAddr {
```

Comment:
Use: 22 bits of address, 2 fixed bits (multicast=no, locally administered=yes), 24 bits of VNI.

Reply:
Fixed in a0cd7a2.